kubernetes: 1.9.7 -> 1.10.3

johanot opened the pull request:
###### Motivation for this change

Update kubernetes module and package to 1.10.X.

This PR is a cleaned and rebased version of latest cherry-picks to https://github.com/NixOS/nixpkgs/pull/38445

This PR also updates kube-dns and kube-dashboard addons to latest versions:

kube-dns: 1.14.4 -> 1.14.10
kube-dashboard: 1.8.2 -> 1.8.3


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [X] other Linux distributions
- [X] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

cstrahan committed:
kubernetes: update to 1.10

Commit:
kubernetes: 1.10.0 -> 1.10.3

azazel75 committed:
Kuberbetes DNS addon 1.14.4 -> 1.14.10

As shipped with k8s 1.10.3.

Also:
- updated the definition jsons as they are distributed in k8s.
- updated the image uris as they are renamed in k8s
- added imageDigest param as per 736848723e5aefa5d24396c58dc6de603399efde

azazel75 committed:
Kuberetes Dashboard addon 1.8.2 -> 1.8.3

As shipped with k8s 1.10.3.

Also:
- updated the definition jsons as they are distributed in k8s.
- updated the image uris as they are renamed in k8s
- added imageDigest param as per 736848723e5aefa5d24396c58dc6de603399efde

azazel75 committed:
Kubernetes tests: increase the size of the VM from 700MB to 1.5GB

VMs were starving, many of the daemons were unable to complete their
tasks resulting in tests failures.

Turned off verbose output from k8s components as it consumes even more resources, and useful error messages actually drown in debug-clutter

srhb commented on pkgs/applications/networking/cluster/kubernetes/default.nix:
```diff
@@ -16,16 +16,16 @@ with lib;
   };
 
-  # go > 1.10 should be fixed by https://github.com/kubernetes/kubernetes/pull/60373
+  # go > 1.10 should be fixed by https://github.com/kubernetes/kubernetes/pull/60597
```
This looks like it's already merged, but I think we should follow whichever version kube is built with upstream, which appears to be 1.9.3. Is there any place where this information is located that we can link to instead? Maybe the changelog?

johanot replied:
According to https://github.com/kubernetes/community/blob/master/contributors/devel/development.md, minimum Golang version required for 1.10 is v1.9.1. And the official version states in External Dependencies for kubernetes 1.10 is v1.9.3: https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG-1.10.md

srhb replied:
Let's link to the External Dependencies instead then.

srhb commented on nixos/modules/services/cluster/kubernetes/default.nix:
```diff
@@ -339,9 +370,9 @@ in {
       };
 
-      admissionControl = mkOption {
+      enableAdmissionPlugins = mkOption {
```
What change does this require a user of admissionControl to make? If any, this should be documented in the release notes for 18.09

srhb commented on nixos/modules/services/cluster/kubernetes/dashboard.nix:
```diff
@@ -113,8 +134,8 @@ in {
           ports = [{
-            port = 80;
-            targetPort = 9090;
+            port = 443;
```
These port changes look like changes to user-facing behaviour, is that right? If so, we need a release notes for it at the very least.

azazel75 replied:
Yes, the ports changed upstream, so I chose to update them  in our manifest

srhb commented:
I've built and run the tests on NixOS. Looks good! :+1:

azazel75 commented:
Well done @johanot, do you need any further help?

coretemp commented:
I am mostly curious whether anyone is using kubernetes on NixOS in production, since the existence of a module is not the same as a production version module.

In general, some kind of meta-data on services like "experimental" or "just a toy" or "used in a national lab" would be useful.

johanot commented:
@coretemp We are using the module for our production cluster. We have a couple of very minor modifications to the module in our setup, which I just haven't gotten around to upstream yet.

@azazel75 Thanks. I'll look into the review comments now and let you know if I need anything further.

johanot commented:
@srhb Please review latest commit, when you have time :)

srhb commented on nixos/doc/manual/release-notes/rl-1809.xml:
```diff
@@ -108,7 +108,7 @@ $ nix-instantiate -E '(import &lt;nixpkgsunstable&gt; {}).gitFull'
      <literal>gnucash</literal> has changed from version 2.4 to 3.x.
      If you've been using <literal>gnucash</literal> (version 2.4) instead of
-     <literal>gnucash26</literal> (version 2.6) you must open your Gnucash 
+     <literal>gnucash26</literal> (version 2.6) you must open your Gnucash
```
No touchy the whitespace of other people! :P

Commit:
Kubernetes:
- Added information regarding breaking changes to release note for 18.09
- Changed golang version comment in kubernetes package
- Added @johanot to maintainers list

srhb commented:
I think this is very good quality and quite sufficient for inclusion. :)

coretemp commented on nixos/modules/services/cluster/kubernetes/dashboard.nix:
```diff
@@ -127,15 +148,56 @@ in {
     } // (optionalAttrs cfg.enableRBAC {
       kubernetes-dashboard-crb = {
-        apiVersion = "rbac.authorization.k8s.io/v1beta1";
+        apiVersion = "rbac.authorization.k8s.io/v1";
```
v1 is repeated in at least two places. Please introduce a variable and use that instead of using the same constant twice.

coretemp commented on nixos/modules/services/cluster/kubernetes/dns.nix:
```diff
@@ -59,7 +59,7 @@ in {
     services.kubernetes.addonManager.addons = {
       kubedns-deployment = {
-        apiVersion = "apps/v1beta1";
+        apiVersion = "extensions/v1beta1";
```
Same comment for a beta version.

coretemp commented on nixos/doc/manual/release-notes/rl-1809.xml:
```diff
@@ -241,6 +241,18 @@ $ nix-instantiate -E '(import &lt;nixpkgsunstable&gt; {}).gitFull'
+   </listitem>
+   <listitem>
+    <para>
+     Recommented way to access the Kubernetes Dashboard is with HTTPS (TLS)
```
- t -> d (must have)
- with -> via (nice to have)

and perhaps specify a complete URL to access the kubernetes dashboard, if possible.(wish list)

johanot replied:
Will fix the typo in an upcoming PR. Thanks :-)

coretemp commented on nixos/modules/services/cluster/kubernetes/dashboard.nix:
```diff
@@ -57,40 +57,61 @@ in {
                 image = "${name}:${version}";
                 ports = [{
-                  containerPort = 9090;
+                  containerPort = 8443;
```
8443 is repeated in multiple places.

coretemp commented on nixos/modules/services/cluster/kubernetes/dns.nix:
```diff
@@ -205,24 +206,24 @@ in {
                   };
+                  args = [
+                    "--v=2"
+                    "--logtostderr"
```
Motivation for logging to stderr is missing. For production services, I almost never want to do that.

azazel75 replied:
It's customnary in Docker containers: logs are then managed using logging drivers and aggregated using dedicated services. The alternative would be for the spawned process to write to files (as there's probably no logging daemon running in the container), but that approach brings up the issue of using a PersistentVolume just for the logs, which is far more invasive as a choice, in my opinion.

johanot replied:
@coretemp No personal preferences involved here. The yaml manifest is a copy of the upstream kube-dns manifest at: https://github.com/kubernetes/kubernetes/blob/release-1.10/cluster/addons/dns/kube-dns.yaml.base. A possible future improvement to the dns submodule could be to add more options for making these things configurable. For now though, I think it's perfectly fine to ship the upstream defaults.

coretemp commented on nixos/modules/services/cluster/kubernetes/dns.nix:
```diff
@@ -205,24 +206,24 @@ in {
+                    failureThreshold = 5;
                   };
+                  args = [
+                    "--v=2"
```
If this is a verbosity level it should have a comment. The numeric value could instead be a named constant saying something about the verbosity level.

johanot commented:
@coretemp Thank you for your comments in general. In a couple of weeks (probably) I'll be submitting a few minor improvements to the kubernetes modules in a new PR. Will bear your considerations in mind then.

NeQuissimus commented:
@johanot ! Somehow I missed this PR completely... Thank you for this! :tada:
